File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if (s2.length ()  < s1.length()) return false;
        
        int[] freq1 = new int[26]; //s1의 빈도수를 저장 a to z: 0 to 25
        int[] freq2 = new int[26]; //s2의 빈도수
        for (int i = 0; i < s1.length(); i++) {
            char c1 = s1.charAt(i);
            freq1[c1 - 'a']++;

            char c2 = s2.charAt(i);
            freq2[c2 - 'a']++;
        }

        //sliding window로 substring찾기

        //초기상태에서 비교
        if (Arrays.equals(freq1, freq2)) return true;

        //진짜 비교
        for (int i = 1; i < s2.length() - s1.length() + 1; i++) {
            //나갈거 나가고 들어올거 나온다음에 비교
            int l = i - 1; //빠질 놈
            freq2[s2.charAt(l) - 'a']--;
            int r = i + s1.length() - 1; //새로 들어올 인덱스
            freq2[s2.charAt(r) - 'a']++;

            if (Arrays.equals(freq1, freq2)) return true;
        }
        return false;
    }
}

File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        freq1 = Counter(s1)
        freq2 = Counter(s2[:len(s1)]) #s1의 길이까지만 저장

        if freq1 == freq2: return True

        for i in range(1, len(s2) - 1 - len(s1) + 1 + 1):
            
            l = i
            r = i + len(s1) - 1
           # print(f"checking {l}, {r}")
            freq2[s2[l-1]] -= 1
            if freq2[s2[l-1]] == 0:
                del freq2[s2[l-1]]
            freq2[s2[r]] += 1

            if freq1 == freq2: return True

        return False
        

'''
s1의 빈도수를 저장한다
s2를 two pointers로 순회한다
    빈도수를 추적하며 빈도수가 같으면 종료한다
    키와 밸류가 모두 동일해야한다.
'''